lib/transaction.c

```diff
@@ -1177,38 +1177,75 @@ struct vfydata_s {
     int vfylevel;
 };
 
+/* order rpmRC codes by severity */
+static int rpmRCseverity(rpmRC rc)
+{
+    switch (rc) {
+    case RPMRC_OK:
+	return 0;
+    case RPMRC_NOTFOUND:
+	return 1;
+    case RPMRC_NOKEY:
+	return 2;
+    case RPMRC_NOTTRUSTED:
+	return 3;
+    case RPMRC_FAIL:
+	return 4;
+    }
+    return rc;
+}
+
 static int vfyCb(struct rpmsinfo_s *sinfo, void *cbdata)
 {
     struct vfydata_s *vd = (struct vfydata_s *)cbdata;
+    int newerror = 0;
+
+
+    int severity = rpmRCseverity(sinfo->rc);
 
     if (sinfo->type & RPMSIG_VERIFIABLE_TYPE && sinfo->rc != RPMRC_NOTFOUND) {
-	int res = (sinfo->rc != RPMRC_OK);
 	/* Take care not to override a previous failure with success */
-	if (res > vd->type[sinfo->type])
-	    vd->type[sinfo->type] = res;
+	if (severity > vd->type[sinfo->type]) {
+	    vd->type[sinfo->type] = severity;
```

Comment:
Don't. Don't abuse a completely unrelated field when there's no reason to do so, just because you think you might get away with it. This is an internal struct, if you need a severity field then add it. And since it can then be used to aid sorting these items, maybe it'll all fall into place more naturally that way. For one, using a nicer data structure (now that we have plenty available) might make it all much saner.

This verification callback fubar was written under RHEL deadline pressure six years ago and isn't exactly one of my pridest moments...

Reply:
Ehm, okay so vd->type is something entirely different (a field badly named by me), I first thought this was abusing the data contained in sinfo->type. This is more harmless but also there's nothing saved by avoiding an extra field for the purpose, this code is bad enough as it is :laughing: